message-format.md

```diff
@@ -0,0 +1,122 @@
+this to the receiver.
+
+If encryption is enabled, a separate encryption header is inserted directly
+after the binary header before the message, a format for this header not yet
+determined
+
+## Binary Header
+```text
+| 0                             | 1                             |
+| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+| MAGIC     | TYPE              | FLAGS                         |
+
+| 2                             | 3                             |
+| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
+| RESERVED: 0x00                | RESERVED: 0x00                |
+
+ MAGIC: 0b101
+ TYPE:  0b00000: Standard Message Type
+ FLAGS: 0: compression: 0: None
+                        1: LZMA
+        1: encryption:  0: No
+                        1: Yes
+```
+
+## Definitions
+
+* **REQUIRED**: this entry must be present in every message
+* **REQUIRED, can be omitted if empty**: the sending node must send the
+  information in this entry if it has any. It does not need to include the entry
+  if it would be emtpy
+* **OPTIONAL**: the sending node can decide if it wishes to send this
+  information i.e. it wishes to activate this feature
+
+## Format
+
+```python
+msg = {
+    # the unique router id, may not contain whitespace or any of the brackets:
+    # []()<>{}
+    # REQUIRED
+    "id": "NODE_ID",
+
+    # strictly monotonically increasing, REQUIRED
+    "seq": 12,
+
+    # Announced networks, OPTIONAL
+    "networks": {
+        "1.2.3.4/12": None,
+        # Additional network information, OPTIONAL
+        "2.3.4.5/23": {
+            "retracted": True
+        }
+    },
+
+    # IP addresses of sending interface, at least one is REQUIRED
+    "addr-v4": "123.45.67.89",
+    "addr-v6": "abcd:ef12::1234",
+
+    # all reachable routes, triggers full replacement in receiver when
+    # present, OPTIONAL
+    "routing-data": {
+        # For every policy
+        "POLICY": {
+            # For every reachable node
+            "TO_NODE_ID": {
+                "path": "NODE_ID>[1]>NODE_ID>[2]>NODE_ID",
+            }
+        }
+    },
+
+    # Only partial upgrades, content not yet defined, OPTIONAL
+    "partial-routing-data": {
+        "tbd": "not yet defined"
+    },
+
+    # all nodes referenced in another section, REQUIRED, can be omitted if empty
+    "node-data": {
+        "NODE_ID": {
+            "networks": {
+                "1.2.3.4/12": None,
+                "2.3.4.5/23": {
+                    "retracted": True
+                }
+            }
+        }
+    },
+
+    # all links referenced in "path" above, REQUIRED, can be omitted if empty
+    "link-attributes": {
+        "1": {
+            "bandwidth": 100,
+            "loss": 0.1
+        },
+        "2": {
+            "bandwidth": 10,
+            "loss": 0.01
+        }
+    },
+
+    # requesting reflection, OPTIONAL
+    "reflector": {
+        "asym-seq": 666  # Example of a reflection request
```

Comment:
I dont like asym-seq, I mean not so bad, but are there any other abbreviation known? Just a minor knit ...

Reply:
Yeah, me neither, I'm still thinking of a good name. Anyway, this is just a example for the later feature of asymmetric neighbor detection, so the name doesn't really matter

Reply:
if we find a nice name we should change it now.

Reply:
question: if a router A announce a reflector one time: how should router B react if this field is never transmitted again. reflect the old, last asym-seq for a lifetime?

what happens to packet loss? is this mechanism clean? I mean in OLSR it is mandatory, so it is far easier. This must be addressed in the ID:

```
If router transmit and request asymmetric detection once via reflector.asym-seq then they MUST set this flag always. Removing this flag in operation is not allowed. 
```

Reply:
The basic idea behind reflection is to give a node feedback what reaches its neighbors. Therefore reflections should contain the newest reflector request it has. If this one was empty/non-existant then so be it.

If a node stops sending reflector requests it should know what it is doing because it should be the only one using this data.

Reply:
No, this is true from a implementation perspective. From a formal protocol specification perspective this MUST be avoided because the behavior is undefined. Think about how the protocol specification should address this?

- "A router MUST never stop sending reflector requests if once started"
- "If a router stop sending reflector requests the reflecting host should send the last request" 
- "If a router stop sending reflector requests the reflecting host MUST stop sending the last request"

Here it is possible to think about how to design a protocol which is extendable and probably used we never thought off. So probably the last bullet point is the best, combined with the first bullet point, lowered to SHOULD.

Reply:
Yeah, you are right, your last suggestion sounds like a good choice.